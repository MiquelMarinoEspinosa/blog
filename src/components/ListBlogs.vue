<template>
  <div id="show-blogs">
    <h1>List Blog Titles</h1>
    <input type="text" v-model="search" placeholder="search blogs" />
    <div
      class="single-blog"
      v-for="(blog, index) in filteredBlogs"
      :key="index"
    >
      <h2>{{ blog.title | uppercase }}</h2>
    </div>
  </div>
</template>

<script>
import SearchMixin from "../mixins/SearchMixin";

export default {
  data() {
    return {
      blogs: [],
      search: "",
    };
  },
  methods: {},
  created() {
    this.$http.get("http://localhost:8081/blog").then((response) => {
      this.blogs = response.data.slice(0, 10);
    });
  },
  computed: {},
  filters: {
    uppercase(value) {
      return value.toUpperCase();
    },
  },
  mixins: [SearchMixin],
};
</script>

<style>
#show-blogs {
  max-width: 800px;
  margin: 0 auto;
}
.single-blog {
  padding: 20px;
  margin: 20px 0;
  box-sizing: border-box;
  background: #eee;
}
</style>
