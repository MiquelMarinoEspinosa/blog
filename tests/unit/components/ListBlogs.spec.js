import { createLocalVue, shallowMount } from "@vue/test-utils";
import Router from "vue-router";
import axios from "axios";
import ListBlogs from "@/components/ListBlogs";

const localVue = createLocalVue();

localVue.use(Router);
jest.mock("axios");

localVue.prototype.$http = axios;

describe("ListBlogs", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders an empty blogs list", async () => {
    const data = [];
    axios.get.mockResolvedValue({ data });
    const listBlogs = shallowMount(ListBlogs, { localVue });
    await localVue.nextTick();
    expect(axios.get).toBeCalledTimes(1);
    expect(axios.get).toBeCalledWith("http://localhost:8081/blog");
    expect(listBlogs.html()).toMatchSnapshot();
  });

  it("search in an empty list returns an empty list blogs", async () => {
    const data = [];
    axios.get.mockResolvedValue({ data });
    const listBlogs = shallowMount(ListBlogs, { localVue });
    await localVue.nextTick();
    const searchBox = listBlogs.get("input[type=text]");
    searchBox.setValue("Blog");
    await localVue.nextTick();
    expect(listBlogs.html()).toMatchSnapshot();
  });

  it("renders a blogs list", async () => {
    const data = [
      {
        title: "Blog 1",
        content: "Blog content 1",
        author: "Author 1",
        categories: ["vue", "react"],
      },
      {
        title: "Blog 2",
        content: "Blog content 2",
        author: "Author 2",
        categories: ["vue"],
      },
      {
        title: "Blog 3",
        content: "Blog content 3",
        author: "Author 2",
        categories: ["react"],
      },
    ];
    axios.get.mockResolvedValue({ data });
    const listBlogs = shallowMount(ListBlogs, { localVue });
    await localVue.nextTick();
    expect(axios.get).toBeCalledTimes(1);
    expect(axios.get).toBeCalledWith("http://localhost:8081/blog");
    expect(listBlogs.html()).toMatchSnapshot();
  });

  it("does not return mode than 10 blog items", async () => {
    const data = [
      {
        title: "Blog 1",
        content: "Blog content 1",
        author: "Author 1",
        categories: ["vue", "react"],
      },
      {
        title: "Blog 2",
        content: "Blog content 2",
        author: "Author 2",
        categories: ["vue"],
      },
      {
        title: "Blog 3",
        content: "Blog content 3",
        author: "Author 2",
        categories: ["react"],
      },
      {
        title: "Blog 4",
        content: "Blog content 4",
        author: "Author 4",
        categories: ["vue", "react"],
      },
      {
        title: "Blog 5",
        content: "Blog content 5",
        author: "Author 5",
        categories: ["vue"],
      },
      {
        title: "Blog 6",
        content: "Blog content 6",
        author: "Author 2",
        categories: ["react"],
      },
      {
        title: "Blog 7",
        content: "Blog content 7",
        author: "Author 4",
        categories: ["vue", "react"],
      },
      {
        title: "Blog 8",
        content: "Blog content 8",
        author: "Author 5",
        categories: ["vue"],
      },
      {
        title: "Blog 9",
        content: "Blog content 9",
        author: "Author 2",
        categories: ["react"],
      },
      {
        title: "Blog 10",
        content: "Blog content 10",
        author: "Author 2",
        categories: ["react"],
      },
      {
        title: "Blog 11",
        content: "Blog content 11",
        author: "Author 2",
        categories: ["react"],
      },
    ];
    axios.get.mockResolvedValue({ data });
    const listBlogs = shallowMount(ListBlogs, { localVue });
    await localVue.nextTick();
    expect(axios.get).toBeCalledTimes(1);
    expect(axios.get).toBeCalledWith("http://localhost:8081/blog");
    expect(listBlogs.html()).toMatchSnapshot();
  });

  it("search and return 2 blogs items", async () => {
    const data = [
      {
        title: "Another Blog 1",
        content: "Blog content 1",
        author: "Author 1",
        categories: ["vue", "react"],
      },
      {
        title: "Another Blog 2",
        content: "Blog content 2",
        author: "Author 2",
        categories: ["vue"],
      },
      {
        title: "Blog 3",
        content: "Blog content 3",
        author: "Author 2",
        categories: ["react"],
      },
    ];
    axios.get.mockResolvedValue({ data });
    const listBlogs = shallowMount(ListBlogs, { localVue });
    await localVue.nextTick();
    const searchBox = listBlogs.get("input[type=text]");
    searchBox.setValue("Another");
    await localVue.nextTick();
    expect(listBlogs.html()).toMatchSnapshot();
  });
});
