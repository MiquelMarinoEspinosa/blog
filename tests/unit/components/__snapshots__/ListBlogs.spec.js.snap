// Jest Snapshot v1, https://goo.gl/fbAQLP
exports[`ListBlogs renders an empty blogs list 1`] = `
<div id="show-blogs">
  <h1>List Blog Titles</h1> <input type="text" placeholder="search blogs">
</div>
`;

exports[`ListBlogs search in an empty list returns an empty list blogs 1`] = `
<div id="show-blogs">
  <h1>List Blog Titles</h1> <input type="text" placeholder="search blogs">
</div>
`;

exports[`ListBlogs renders a blogs list 1`] = `
<div id="show-blogs">
  <h1>List Blog Titles</h1> <input type="text" placeholder="search blogs">
  <div class="single-blog">
    <h2>BLOG 1</h2>
  </div>
  <div class="single-blog">
    <h2>BLOG 2</h2>
  </div>
  <div class="single-blog">
    <h2>BLOG 3</h2>
  </div>
</div>
`;


exports[`ListBlogs search and return 2 blogs items 1`] = `
<div id="show-blogs">
  <h1>List Blog Titles</h1> <input type="text" placeholder="search blogs">
  <div class="single-blog">
    <h2>ANOTHER BLOG 1</h2>
  </div>
  <div class="single-blog">
    <h2>ANOTHER BLOG 2</h2>
  </div>
</div>
`;

exports[`ListBlogs does not return mode than 10 blog items 1`] = `
<div id="show-blogs">
  <h1>List Blog Titles</h1> <input type="text" placeholder="search blogs">
  <div class="single-blog">
    <h2>BLOG 1</h2>
  </div>
  <div class="single-blog">
    <h2>BLOG 2</h2>
  </div>
  <div class="single-blog">
    <h2>BLOG 3</h2>
  </div>
  <div class="single-blog">
    <h2>BLOG 4</h2>
  </div>
  <div class="single-blog">
    <h2>BLOG 5</h2>
  </div>
  <div class="single-blog">
    <h2>BLOG 6</h2>
  </div>
  <div class="single-blog">
    <h2>BLOG 7</h2>
  </div>
  <div class="single-blog">
    <h2>BLOG 8</h2>
  </div>
  <div class="single-blog">
    <h2>BLOG 9</h2>
  </div>
  <div class="single-blog">
    <h2>BLOG 10</h2>
  </div>
</div>
`;
